Add inline GLSL, don't use `proc_macro_hack`, update dependencies

wackbyte opened the pull request:
Inline GLSL is written like this:
```rs
inline_glsl! {
    version: 450, kind: vert, optimize: size, target: vulkan1_1,
    r#"
#version 450

void main() {
    gl_Position = vec4(0);
}
"#
};
```
I could probably change it to use a comma instead of a semicolon after the compile options, if you prefer.

I also merged the impl crate since `proc_macro_hack` is no longer needed.

(Closes #3)

wackbyte committed:
Update dependencies and don't use proc_macro_hack (remove impl crate)

Ralith commented:
Thanks, this looks like a great improvement! Could you split the second commit to separate out the code reorganization with no intended behavior changes from the introduction of a new macro? That would make review much easier.

Also, thoughts on naming the inline macro `glsl!` rather than `inline_glsl!`?

wackbyte commented:
Sorry, I'm confused on what you mean by splitting the commit, could you please explain?
And I do like `glsl!` better, so I'll add that.

Ralith commented:
Replacing the middle commit with two separate commits, one of which only reorganizes the code, and the other of which only introduces new behavior.

wackbyte committed:
Refactor codegen, test includes

wackbyte committed:
Support inline GLSL

wackbyte commented:
Done!

Ralith commented on src/lib.rs:
````diff
@@ -69,3 +91,43 @@ pub fn include_glsl(tokens: TokenStream) -> TokenStream {
+///
+///   ```glsl
+///   // standard
+///   #include <crate/to/include.glsl>
````
This example is confusing, since it suggests that the crate itself must be part of the path. Maybe drop the code example and just specify that includes will always be resolved relative to the crate root?

Ralith commented on src/lib.rs:
```diff
@@ -69,3 +91,43 @@ pub fn include_glsl(tokens: TokenStream) -> TokenStream {
+/// # Limitations
+///
+/// - Because the shader kind cannot be inferred from a file extension,
+///   you may need to specify it manually as the above example does.
```
e.g. `#pragma shader_stage(vertex)` in the GLSL should also work.

wackbyte committed:
Allow glsl! to be invoked with no options and don't use a semicolon

Also allow a trailing comma after the source string

This also allows the comma after the options in glsl! to be omitted,
but I don't think this is an issue.

wackbyte committed:
Improve glsl! documentation

Ralith commented on src/build.rs:
```diff
@@ -56,69 +56,77 @@ impl Default for BuildOptions {
-                        out.kind = Some(kind);
-                    } else {
-                        return Err(syn::Error::new(value.span(), "unknown shader kind"));
+            if input.peek(Ident) {
```
Style nit: we can avoid the rightward drift here with `if !input.peek(Ident) { break; }` or similar.

wackbyte replied:
Fixed it by replacing the `while` condition to do the equivalent

wackbyte committed:
Fix rightward drift and require comma after options in glsl!

Ralith reviewed (approved):
Thanks! Merged manually.

Ralith commented:
Released in 0.2.8.
